### ISSUE TITLE
Edge should be a Generic type so schema can be inferred from it.

### ISSUE THREAD

**Issue by bpatters** [graphql-java/graphql-java#78]
I'm currently working on a Java POJO --> GraphQL schema generator (see github schemagen-graphql) and the EdgeClass needs to be a Generic so that node type can be inferred during schema generation.

**Comment by andimarek** [comment]
Hi,

could you please be more specific and maybe post a example?

Thanks

**Comment by bpatters** [comment]
Yea here's my local repository version of the class:
https://github.com/bpatters/schemagen-graphql/blob/master/src/main/java/com/bretpatterson/schemagen/graphql/relay/Edge.java

And my local version of Connection.java class
https://github.com/bpatters/schemagen-graphql/blob/master/src/main/java/com/bretpatterson/schemagen/graphql/relay/RelayConnection.java


These are both Generics so that I can use reflection to generate the GraphQL schema from the definition of the object. For example in my test case I have a:
GameDTO which has a connection to the users in the game. I need to define this as:
Connection<UserDTO> and then have the type propagate down through the Connection object.

IE:
```
public class Connection<T> {
	private List<Edge<T>> edges;
	private PageInfo pageInfo;
}
```

this way I can define the Connection as above, and the schemgen generator can fully qualify the GraphQL schema based on the definition of:
```
public class GameDTO {
     Connection<UserDTO> users;
}
```
So that the GraphQL Schema generation can fully generate the schema from the definition.

**Comment by bpatters** [comment]
I'm currently working on get my schemagen-graphql production ready for a large enterprise app in which I'm hoping to use GraphQL + Relay for, but will fall back to just GraphQL if Relay works out to not be a good fit. So let me give you more information in the context of a real world usage scenario

I've got many top level queries that return Lists of objects (And objects containing lists of other objects). For example
```java
List<AssetDTO>  findAssets(... filtering/paging parameters);
List<AbstractCalendarItemDTO> findCalendarItems(... filter/paging parameters);
```

Both of these endpoints I'm converting to be Connection objects. What I'd like to do is use a Single RelayConnection Object and a single Edge object, since they're really just container objects and specializations would increase maintenance with no real value. However; I also want the Schema to be strongly typed so the UI team can browse the schema (GraphIQL is great!) and know what Objects are returned and their fields etc.

The interface I've implemented, and still working through details as I learn more, is:
(These are actually Wrapped in top level Objects since Relay doesn't support multiple parameters nor object parameters for top level queries, though graphql itself does).
```java
RelayConnection<AssetDTO>  findAssets(... filtering/paging parameters);
RelayConnection<AbstractCalendarItemDTO> findCalendarItems(... filter/paging parameters);
```

Where RelayConnection is a simple Generic container around an Edge and PageInfo object. I pass down the Type Parameter to the Edge so that the Edge's Node object can be Strongly typed also. This looks like:
```java
public class RelayConnection<T> {
	private List<Edge<T>> edges;
	private PageInfo pageInfo;
...
}

public class Edge<T> {
	T node;
	ConnectionCursor cursor;
...
```

This structure allows these two objects to be used for numerous types of connections and edges while maintaining strong typing. At Schema Generation time I traverse the Generics for all method signatures and POJO object signatures to generate a strongly typed GraphQL schema.
So the above actually turns into the following Object Types:
```
RelayConnection_AssetDTO
Edge_AssetDTO
RelayConnection_AbstractCalendarItem
Edge_AbstractCalendarItem
```
Here I've replaced the Generic Type with fully qualified Type names using the Pattern of Class_TypeArg1_TypeArg2_..TypeArg*N*


I'm still in the middle stages of development, so I'm still learning the finer points of Java Types with Generics ---> GraphQL schema definitions so it's quite possible I'm missing something. I'm learning a lot everyday. I also don't strictly need these objects changed, I can use my own internal ones, however I'd like to not cause confusing between my library and this library. As this GraphQL-java library is really well done and is the foundation I'm building upon by making Schema Generation and Connection between the Schema and Implementation declarative with minimal annotations.

Let me know if you want more information or a PR. I'm happy to contribute. Excellent product btw.
